plugins/tiletypes.cpp

```diff
@@ -1097,54 +1222,119 @@ command_result df_tiletypes_here_point (color_ostream &out, vector <string> & pa
+        out.printerr("Invalid hidden value: %d\n", target.hidden);
+        return false;
+    }
+    if (target.light < -1 || target.light > 1) {
+        out.printerr("Invalid light value: %d\n", target.light);
+        return false;
+    }
+    if (target.subterranean < -1 || target.subterranean > 1) {
+        out.printerr("Invalid subterranean value: %d\n", target.subterranean);
+        return false;
+    }
+    if (target.skyview < -1 || target.skyview > 1) {
+        out.printerr("Invalid skyview value: %d\n", target.skyview);
+        return false;
+    }
+    if (target.aquifer < -1 || target.aquifer > 2) {
+        out.printerr("Invalid aquifer value: %d\n", target.aquifer);
+        return false;
+    }
+    if (target.material == df::tiletype_material::STONE) {
+        if (!isStoneInorganic(target.stone_material)) {
+            out.printerr("Invalid stone material: %d\n", target.stone_material);
+            return false;
+        }
+        if (!is_valid_enum_item(target.vein_type)) {
+            out.printerr("Invalid vein type: %d\n", target.vein_type);
+            return false;
+        }
+    }
+    else {
+        target.stone_material = -1;
+        target.vein_type = df::inclusion_type::CLUSTER;
+    }
 
     MapExtras::MapCache map;
-    return paintTile(map, pos, target) && map.WriteAll();
+    PaintResult result = paintTile(map, pos, target);
+    if (result.paintCount > 0 && map.WriteAll()) {
+        result.postWrite();
+        return true;
+    }
+    return false;
 }
 
 static int tiletypes_setTile(lua_State *L) {
     color_ostream *out = Lua::GetOutput(L);
     if (!out)
         out = &Core::getInstance().getConsole();
 
+    TileType target = TileType();
+
     df::coord pos;
     Lua::CheckDFAssign(L, &pos, 1);
-    df::tiletype_shape shape = (df::tiletype_shape)lua_tointeger(L, 2);
-    df::tiletype_material material = (df::tiletype_material)lua_tointeger(L, 3);
-    df::tiletype_special special = (df::tiletype_special)lua_tointeger(L, 4);
-    df::tiletype_variant variant = (df::tiletype_variant)lua_tointeger(L, 5);
+    if (lua_istable(L, 2)) {
+        auto lua_getintfield = [L](const char* fieldName, int defaultValue) -> int {
+            lua_getfield(L, 2, fieldName);
+            int output = defaultValue;
+            if (!lua_isnil(L, -1)) {
+                output = lua_tointeger(L, -1);
+            }
+            lua_pop(L, 1);
+            return output;
+        };
+
+        target.shape     = (df::tiletype_shape)lua_getintfield("shape", target.shape);
+        target.material  = (df::tiletype_material)lua_getintfield("material", target.material);
+        target.special   = (df::tiletype_special)lua_getintfield("special", target.special);
+        target.variant   = (df::tiletype_variant)lua_getintfield("variant", target.variant);
+        target.dig            = lua_getintfield("dig", target.dig);
+        target.hidden         = lua_getintfield("hidden", target.hidden);
```

Comment:
these would normally be specified as booleans. does that work correctly with this logic?

Reply:
Those were kept as integers to match up with how tiletypes works from the command line, where you specify -1, 0, or 1 for those options. It could be changed to nil, false, or true for Lua, but it didn't seem worth it to introduce such an inconsistency, especially since aquifer has to be an integer now that it accepts -1, 0, 1, or 2.

Reply:
that's fine. I'd document that, though.